### PR TITLE
Core: Create a SuppressibleExtension to provide the Suppressed property to all the WBs

### LINKED ISSUE #12361: [Problem] Suppressed objects are not distinguishable
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

#12096 introduce a Suppressed property for PartDesign features and #12359 aims to add it for FEM constraints.
The tree items should reflect the suppressed status i.e. the tree item text should be striked through.

### Full version info

```shell
OS: Debian GNU/Linux trixie/sid (XFCE/xfce)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.35932 +3 (Git)
Build type: Release
Branch: fem_suppress
Hash: 33634517a7a63a8b0b051a1e62f41de2a8ae2c51
Python 3.11.8, Qt 5.15.10, Coin 4.0.2, Vtk 9.1.0, OCC 7.6.3
Locale: French/France (fr_FR)
```

### Subproject(s) affected?

PartDesign

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by FlachyJoe**
Many objects can be suppressed, this extension aims to provide a unique visualization in the tree view.

Fix #12361

### CODE REVIEW

**Comment by FlachyJoe** [comment]
@wwmayer could you please take a look? ViewProviderSuppressibleExtension::extensionUpdateData is never called.

**Comment by FlachyJoe** [comment]
image

**Comment by FlachyJoe** [comment]
Ready for review.
I have no possibility to debug MacOS nor Windows build, it certainly lacks one new file's mention in a config.

**Comment by FlachyJoe** [comment]
Back to draft: a crash occurs at object deletion.

**Comment by FlachyJoe** [comment]
Got it. I was confused by
```
>>> doc=App.ActiveDocument
>>> doc.Body.addExtension('App::SuppressibleExtension')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: Extension is not a python addable version: 'App::SuppressibleExtension'
```

But using `App::SuppressibleExtensionPython` is ok
```
>>> doc.Body.addExtension('App::SuppressibleExtensionPython')
>>> doc.Body.ViewObject.addExtension('Gui::ViewProviderSuppressibleExtensionPython')
>>> 
```
image

**Comment by wwmayer** [comment]
In how far is this related to https://github.com/FreeCAD/FreeCAD/issues/5545?
Original ticket in Mantis: https://tracker.freecad.org/view.php?id=402
Forum thread: https://forum.freecad.org/viewtopic.php?t=1070

**Comment by FlachyJoe** [comment]
@wwmayer  I don't think a Suppressed object aims to be a Emancipated one (cut from parents).

**Comment by FlachyJoe** [comment]
I've added a Suppress/Unsuppress action in the tree context menu. This action is only shown if the Suppressed property of the object is not hidden so the action visibility is managed at workbench level.
See https://github.com/FlachyJoe/FreeCAD/blob/7b90aea42915c87eaa643c96fe0758b17727a1e6/src/Mod/PartDesign/App/Feature.cpp#L63
```cpp
    App::SuppressibleExtension::initExtension(this);
    Suppressed.setStatus(App::Property::Status::Hidden, true); //Todo: remove when TNP fixed
```

**Comment by PaddleStroke** [comment]
Master does not build today coming from this PR

> Severity	Code	Description	Project	File	Line	Suppression State
> Error	C2027	use of undefined type 'Gui::Document'	FreeCADGui	C:\Users\Sky_l\Desktop\Freecad-Src\src\Gui\ViewProviderSuppressibleExtension.cpp	59	
> Error	C2039	'ActionFunction': is not a member of 'Gui'	FreeCADGui	C:\Users\Sky_l\Desktop\Freecad-Src\src\Gui\ViewProviderSuppressibleExtension.cpp	110	
> Error	C2065	'ActionFunction': undeclared identifier	FreeCADGui	C:\Users\Sky_l\Desktop\Freecad-Src\src\Gui\ViewProviderSuppressibleExtension.cpp	110	
> Error	C2065	'func': undeclared identifier	FreeCADGui	C:\Users\Sky_l\Desktop\Freecad-Src\src\Gui\ViewProviderSuppressibleExtension.cpp	110	
> Error	C2039	'ActionFunction': is not a member of 'Gui'	FreeCADGui	C:\Users\Sky_l\Desktop\Freecad-Src\src\Gui\ViewProviderSuppressibleExtension.cpp	110	
> Error	C2061	syntax error: identifier 'ActionFunction'	FreeCADGui	C:\Users\Sky_l\Desktop\Freecad-Src\src\Gui\ViewProviderSuppressibleExtension.cpp	110	
> Error	C2065	'func': undeclared identifier	FreeCADGui	C:\Users\Sky_l\Desktop\Freecad-Src\src\Gui\ViewProviderSuppressibleExtension.cpp	117	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderPrimitive.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderMirror.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderPython.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderBoolean.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderRuledSurface.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderImport.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderExtrusion.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderScale.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProvider2DObject.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderRegularPolygon.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderTorusParametric.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderCurveNet.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderSpline.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderSphereParametric.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderCylinderParametric.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderConeParametric.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderPrism.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderPointParametric.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderEllipseParametric.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderHelixParametric.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderPlaneParametric.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderBox.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderCompound.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderCircleParametric.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderLineParametric.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\CrossSections.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProvider.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderExt.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	PartGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Part\Gui\ViewProviderReference.obj	1	
> Error	LNK1120	1 unresolved externals	PartGui	C:\Users\Sky_l\Desktop\FC build\Mod\Part\PartGui.pyd	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	SketcherGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Sketcher\Gui\ViewProviderSketch.obj	1	
> Error	LNK2001	unresolved external symbol "public: virtual bool __cdecl Gui::ViewProvider::acceptReorderingObjects(void)const " (?acceptReorderingObjects@ViewProvider@Gui@@UEBA_NXZ)	SketcherGui	C:\Users\Sky_l\Desktop\FC build\src\Mod\Sketcher\Gui\ViewProviderPython.obj	1

File: src/Gui/ViewProviderSuppressibleExtension.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+#ifndef VIEWPROVIDERSUPPRESSIBLEEXTENSION_H
+#define VIEWPROVIDERSUPPRESSIBLEEXTENSION_H
+
+#include "ViewProviderExtensionPython.h"
+
+
+namespace Gui
+{
+
+class ViewProviderSuppressibleExtension : public Gui::ViewProviderExtension
```
Comment (PaddleStroke) [inline src/Gui/ViewProviderSuppressibleExtension.h:32]: This line should be : 
`class GuiExport ViewProviderSuppressibleExtension : public ViewProviderExtension`

File: src/Gui/ViewProviderSuppressibleExtension.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    void extensionUpdateData(const App::Property* prop) override;
+
+    void setSuppressedIcon(bool onoff);
+    QIcon extensionMergeColorfullOverlayIcons (const QIcon & orig) const override;
+    void extensionSetupContextMenu(QMenu* menu, QObject*, const char*) override;
+
+private:
+    bool isSetSuppressedIcon{false};
+};
+
```
Comment (PaddleStroke) [inline src/Gui/ViewProviderSuppressibleExtension.h:49]: `namespace Gui` should start here. See ViewProviderGroupExtension as example

File: src/Gui/ViewProviderSuppressibleExtension.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "ActionFunction.h"
+#include "Control.h"
+#include "Document.h"
+#include "PreCompiled.h"
```
Comment (PaddleStroke) [inline src/Gui/ViewProviderSuppressibleExtension.cpp:26]: "PreCompiled.h" must be the first...
Building now

**Comment by FlachyJoe** [comment]
Thank you @PaddleStroke for the fix.
Isn't the CI supposed to fail before the merge?

**Comment by PaddleStroke** [comment]
The CI is not using the precompiled headers. This is something that is being discussed with @adrianinsaval and @chennes if I recall correctly

**Comment by FEA-eng** [comment]
@FlachyJoe Is it still in development ? I can't find the Suppress option in Properties or in the tree context menu in Part Design.

```
OS: Windows 10 build 19045
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.36277 (Git)
Build type: Release
Branch: main
Hash: 9e1903d46112b3660bf10c6a4537d728101d560b
Python 3.10.13, Qt 5.15.8, Coin 4.0.2, Vtk 9.2.6, OCC 7.6.3
Locale: Polish/Poland (pl_PL)
```

**Comment by FlachyJoe** [comment]
@FEA-eng in PD the Suppress property is hidden (right-click, show all) until the TNP fix is merged due to the problems it can produce in shapes hierarchy.

**Pull request merged**